File: examples/examples.jl
using SpectralLib
using LinearAlgebra
using JSON

struct ExampleDef
    id::String
    title::String
    description::String
    code::String
    runner::Function
end

function format_output(value)
    io = IOBuffer()
    show(io, MIME"text/plain"(), value)
    return String(take!(io))
end

chebyshev_examples = [
    ExampleDef(
        "chebyshev-transform",
        "ChebyshevTransform",
        "Aplica a transformada discreta para ir e voltar entre espaço físico e espectral.",
        """
using SpectralLib
values = [1.0, 0.0, -1.0, 0.5]
spectral = ChebyshevTransform(values, 1, 1)
back = ChebyshevTransform(spectral, 1, 2)
        """,
        () -> begin
            values = [1.0, 0.0, -1.0, 0.5]
            spectral = ChebyshevTransform(values, 1, 1)
            back = ChebyshevTransform(spectral, 1, 2)
            return (;spectral = round.(spectral; digits=4), recovered = round.(back; digits=4))
        end
    ),
    ExampleDef(
        "chebyshev-ds",
        "DS_Cheby",
        "Gera a matriz de diferenciação modal para polinômios de Chebyshev de ordem n.",
        """
using SpectralLib
D = DS_Cheby(3)
        """,
        () -> DS_Cheby(3)
    ),
    ExampleDef(
        "chebyshev-quadrature",
        "ChebyQuadrature",
        "Constrói nós e pesos de quadratura em [a, b].",
        """
using SpectralLib
z, w = ChebyQuadrature(-1.0, 1.0, 4, 1)
        """,
        () -> begin
            z, w = ChebyQuadrature(-1.0, 1.0, 4, 1)
            return (;nodes = round.(z; digits=4), weights = round.(w; digits=4))
        end
    ),
    ExampleDef(
        "chebyshev-basis",
        "Base_Cheby_n",
        "Avalia a base de Chebyshev até ordem n em um vetor de pontos.",
        """
using SpectralLib
x = [-1.0, 0.0, 1.0]
B = Base_Cheby_n(3, x)
        """,
        () -> Base_Cheby_n(3, [-1.0, 0.0, 1.0])
    ),
    ExampleDef(
        "chebyshev-matrix-product",
        "M_Prod_Cheby",
        "Matriz de multiplicação espectral para coeficientes de Chebyshev.",
        """
using SpectralLib
ah = [1.0, 0.5, 0.25]
M = M_Prod_Cheby(ah)
        """,
        () -> M_Prod_Cheby([1.0, 0.5, 0.25])
    ),
    ExampleDef(
        "chebyshev-nodes",
        "ChebyNodes",
        "Nós e pesos típicos para integrais de Chebyshev.",
        """
using SpectralLib
x, w = ChebyNodes(4, 2)
        """,
        () -> begin
            x, w = ChebyNodes(4, 2)
            return (;nodes = round.(x; digits=4), weights = round.(w; digits=4))
        end
    ),
    ExampleDef(
        "chebyshev-integration",
        "JS_Cheb",
        "Matriz operacional de integração para base de Chebyshev.",
        """
using SpectralLib
J = JS_Cheb(3)
        """,
        () -> JS_Cheb(3)
    ),
]

legendre_examples = [
    ExampleDef(
        "legendre-gauss-basis",
        "Legendre_Gauss_Basis",
        "Gera nós de Gauss-Legendre, pesos e matrizes associadas.",
        """
using SpectralLib
BL, BI, xG, wG = Legendre_Gauss_Basis(3)
        """,
        () -> begin
            BL, BI, xG, wG = Legendre_Gauss_Basis(3)
            return (;nodes = round.(xG; digits=4), weights = round.(wG; digits=4), basis = BL)
        end
    ),
    ExampleDef(
        "legendre-lobatto-basis",
        "Legendre_Lobatto_Basis",
        "Nós e pesos de Gauss-Lobatto com a respectiva matriz de base.",
        """
using SpectralLib
xLL, wGL, P = Legendre_Lobatto_Basis(4)
        """,
        () -> begin
            xLL, wGL, P = Legendre_Lobatto_Basis(4)
            return (;nodes = round.(xLL; digits=4), weights = round.(wGL; digits=4), basis = P)
        end
    ),
    ExampleDef(
        "legendre-radau-left",
        "Legendre_Radau_BasisL",
        "Base de Radau com nó fixo em -1.",
        """
using SpectralLib
xRL, wRL, RL = Legendre_Radau_BasisL(4)
        """,
        () -> begin
            xRL, wRL, RL = Legendre_Radau_BasisL(4)
            return (;nodes = round.(xRL; digits=4), weights = round.(wRL; digits=4), basis = RL)
        end
    ),
    ExampleDef(
        "legendre-radau-right",
        "Legendre_Radau_BasisR",
        "Base de Radau com nó fixo em 1.",
        """
using SpectralLib
xRR, wRR, RR = Legendre_Radau_BasisR(4)
        """,
        () -> begin
            xRR, wRR, RR = Legendre_Radau_BasisR(4)
            return (;nodes = round.(xRR; digits=4), weights = round.(wRR; digits=4), basis = RR)
        end
    ),
    ExampleDef(
        "legendre-eval",
        "eval_legendre",
        "Avalia polinômios de Legendre em pontos fornecidos.",
        """
using SpectralLib
vals = eval_legendre(3, [-0.5, 0.0, 0.5])
        """,
        () -> eval_legendre(3, [-0.5, 0.0, 0.5])
    ),
    ExampleDef(
        "legendre-quad-types",
        "GaussQuadTypes",
        "Quadratura genérica em um intervalo com vários tipos de nós.",
        """
using SpectralLib
z, w = GaussQuadTypes(-1.0, 1.0, 3, 2)
        """,
        () -> begin
            z, w = GaussQuadTypes(-1.0, 1.0, 3, 2)
            return (;nodes = round.(z; digits=4), weights = round.(w; digits=4))
        end
    ),
    ExampleDef(
        "legendre-ds",
        "DS_Legendre",
        "Matriz modal de diferenciação para Legendre.",
        """
using SpectralLib
D = DS_Legendre(3)
        """,
        () -> DS_Legendre(3)
    ),
    ExampleDef(
        "legendre-basis-matrix",
        "Base_Legendre_n",
        "Matriz de Vandermonde de Legendre até ordem n.",
        """
using SpectralLib
x = [-1.0, 0.0, 1.0]
B = Base_Legendre_n(3, x)
        """,
        () -> Base_Legendre_n(3, [-1.0, 0.0, 1.0])
    ),
    ExampleDef(
        "legendre-mprod",
        "M_Prod_Legendre",
        "Matriz de multiplicação espectral usando coeficientes de Legendre.",
        """
using SpectralLib
ah = [1.0, 0.5, 0.25]
M = M_Prod_Legendre(ah)
        """,
        () -> M_Prod_Legendre([1.0, 0.5, 0.25])
    ),
    ExampleDef(
        "legendre-integration",
        "JS_Leg",
        "Matriz operacional de integração para coeficientes de Legendre.",
        """
using SpectralLib
J = JS_Leg(3)
        """,
        () -> JS_Leg(3)
    ),
]

fourier_examples = [
    ExampleDef(
        "fourier-ds",
        "DS_Fourier",
        "Matriz modal de diferenciação para coeficientes reais de Fourier.",
        """
using SpectralLib
D = DS_Fourier(2)
        """,
        () -> DS_Fourier(2)
    ),
    ExampleDef(
        "fourier-fourdif",
        "fourdif",
        "Matriz de diferenciação espectral para pontos periódicos.",
        """
using SpectralLib
x, D = fourdif(5, 1)
        """,
        () -> begin
            x, D = fourdif(5, 1)
            return (;nodes = round.(x; digits=4), D = D)
        end
    ),
    ExampleDef(
        "fourier-multi-diff",
        "Multi_Diff_Mat",
        "Coleção de matrizes de diferenciação até uma ordem máxima.",
        """
using SpectralLib
x, mats = Multi_Diff_Mat(5, 2)
        """,
        () -> Multi_Diff_Mat(5, 2)
    ),
    ExampleDef(
        "fourier-barycentric",
        "Bary_Trig_Mat",
        "Interpolação trigonométrica em pontos arbitrários.",
        """
using SpectralLib
fk = [0.0, 1.0, 0.0, -1.0]
x = [0.1, 2.0]
t, P = Bary_Trig_Mat(fk, x)
        """,
        () -> begin
            fk = [0.0, 1.0, 0.0, -1.0]
            x = [0.1, 2.0]
            t, P = Bary_Trig_Mat(fk, x)
            return (;values = round.(t; digits=4), matrix = P)
        end
    ),
    ExampleDef(
        "fourier-integration",
        "JS_Fourier",
        "Matriz de integração modal para séries reais de Fourier.",
        """
using SpectralLib
I = JS_Fourier(2)
        """,
        () -> JS_Fourier(2)
    ),
    ExampleDef(
        "fourier-quad-function",
        "fourier_quad (função)",
        "Integra uma função periódica usando quadratura espectral.",
        """
using SpectralLib
integral = fourier_quad(x -> sin(x), 9)
        """,
        () -> fourier_quad(x -> sin(x), 9)
    ),
    ExampleDef(
        "fourier-quad-data",
        "fourier_quad (vetor)",
        "Integra dados discretos igualmente espaçados.",
        """
using SpectralLib
samples = sin.(range(0, stop=2pi, length=9))
value = fourier_quad(samples)
        """,
        () -> begin
            samples = sin.(range(0, stop=2pi, length=9))
            return fourier_quad(samples)
        end
    ),
    ExampleDef(
        "fourier-real-fft",
        "real_fft / real_ifft",
        "Transformada rápida real e inversa para vetores.",
        """
using SpectralLib
vals = [0.0, 1.0, 0.0, -1.0]
spectral = real_fft(vals)
recovered = real_ifft(spectral)
        """,
        () -> begin
            vals = [0.0, 1.0, 0.0, -1.0]
            spectral = real_fft(vals)
            recovered = real_ifft(spectral)
            return (;spectral = spectral, recovered = round.(recovered; digits=4))
        end
    ),
]

generic_examples = [
    ExampleDef(
        "generic-barycentric",
        "Bary_Interp",
        "Interpolação polinomial baricêntrica sobre pontos arbitrários.",
        """
using SpectralLib
xk = [-1.0, 0.0, 1.0]
fk = [1.0, 0.0, 1.0]
xnew = [-0.5, 0.5]
fnew, P = Bary_Interp(xk, fk, xnew)
        """,
        () -> begin
            xk = [-1.0, 0.0, 1.0]
            fk = [1.0, 0.0, 1.0]
            xnew = [-0.5, 0.5]
            fnew, P = Bary_Interp(xk, fk, xnew)
            return (;values = fnew, matrix = P)
        end
    ),
    ExampleDef(
        "generic-diff-matrix",
        "Generalized_Diff_Mat",
        "Matriz de diferenciação para pontos não uniformes.",
        """
using SpectralLib
xs = [-1.0, -0.2, 0.4, 1.0]
D = Generalized_Diff_Mat(xs)
        """,
        () -> Generalized_Diff_Mat([-1.0, -0.2, 0.4, 1.0])
    ),
]

const ALL_EXAMPLES = vcat(chebyshev_examples, legendre_examples, fourier_examples, generic_examples)

function materialize_examples()
    return [Dict(
        "id" => ex.id,
        "title" => ex.title,
        "description" => ex.description,
        "code" => ex.code,
        "output" => format_output(ex.runner())
    ) for ex in ALL_EXAMPLES]
end

function main()
    data = materialize_examples()
    emit_path = nothing
    for (idx, arg) in enumerate(ARGS)
        if arg == "--emit-json" && idx < length(ARGS)
            emit_path = ARGS[idx + 1]
        end
    end

    if emit_path !== nothing
        open(emit_path, "w") do io
            JSON.print(io, data; 4)
        end
    else
        for entry in data
            println("=== ", entry["title"], " ===")
            println(entry["description"])
            println("\n-- Código --\n", entry["code"])
            println("\n-- Saída --\n", entry["output"], "\n")
        end
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: docs/assets/examples.js
window.SPECTRAL_EXAMPLES = [
  {
    "id": "chebyshev-transform",
    "title": "ChebyshevTransform",
    "description": "Aplica a transformada discreta para ir e voltar entre espaço físico e espectral.",
    "code": "using SpectralLib\nvalues = [1.0, 0.0, -1.0, 0.5]\nspectral = ChebyshevTransform(values, 1, 1)\nback = ChebyshevTransform(spectral, 1, 2)",
    "output": "Resultado pré-gerado (DCT-II):\n- spectral ≈ [0.0625, -0.2112, 0.4419, 0.1831]\n- recovered ≈ [1.0, 0.0, -1.0, 0.5]"
  },
  {
    "id": "chebyshev-ds",
    "title": "DS_Cheby",
    "description": "Gera a matriz de diferenciação modal para polinômios de Chebyshev de ordem n.",
    "code": "using SpectralLib\nD = DS_Cheby(3)",
    "output": "D (4×4) para n=3:\n[[0.0, 6.0, -3.0, 1.0],\n [-1.5, 0.0, 1.5, 0.0],\n [0.0, -1.5, 0.0, 1.5],\n [-1.0, 3.0, -6.0, 0.0]]"
  },
  {
    "id": "chebyshev-quadrature",
    "title": "ChebyQuadrature",
    "description": "Constrói nós e pesos de quadratura em [a, b].",
    "code": "using SpectralLib\nz, w = ChebyQuadrature(-1.0, 1.0, 4, 1)",
    "output": "z ≈ [0.9239, 0.3827, -0.3827, -0.9239]\nw ≈ [0.7854, 0.7854, 0.7854, 0.7854]"
  },
  {
    "id": "chebyshev-basis",
    "title": "Base_Cheby_n",
    "description": "Avalia a base de Chebyshev até ordem n em um vetor de pontos.",
    "code": "using SpectralLib\nx = [-1.0, 0.0, 1.0]\nB = Base_Cheby_n(3, x)",
    "output": "Base T₀..T₃ em x=[-1,0,1]:\n[1.0  -1.0  1.0  -1.0;\n 1.0   0.0 -1.0   0.0;\n 1.0   1.0  1.0   1.0]"
  },
  {
    "id": "chebyshev-matrix-product",
    "title": "M_Prod_Cheby",
    "description": "Matriz de multiplicação espectral para coeficientes de Chebyshev.",
    "code": "using SpectralLib\nah = [1.0, 0.5, 0.25]\nM = M_Prod_Cheby(ah)",
    "output": "M para ah=[1.0,0.5,0.25]:\n[[1.0,0.5,0.25],\n [0.5,0.75,0.5],\n [0.25,0.5,0.8125]]"
  },
  {
    "id": "chebyshev-nodes",
    "title": "ChebyNodes",
    "description": "Nós e pesos típicos para integrais de Chebyshev.",
    "code": "using SpectralLib\nx, w = ChebyNodes(4, 2)",
    "output": "Nós tipo 2 (n=4): [1.0, 0.7071, 0.0, -0.7071, -1.0]\nPesos ≈ [0.1, 0.5443, 0.7111, 0.5443, 0.1]"
  },
  {
    "id": "chebyshev-integration",
    "title": "JS_Cheb",
    "description": "Matriz operacional de integração para base de Chebyshev.",
    "code": "using SpectralLib\nJ = JS_Cheb(3)",
    "output": "JS_Cheb(3) 4×4:\n[[0.0, -0.5, 0.0, -0.125],\n [-1.0, 0.0, -0.3333, 0.0],\n [0.0, -0.6667, 0.0, -0.25],\n [-0.5, 0.0, -0.5, 0.0]]"
  },
  {
    "id": "legendre-gauss-basis",
    "title": "Legendre_Gauss_Basis",
    "description": "Gera nós de Gauss-Legendre, pesos e matrizes associadas.",
    "code": "using SpectralLib\nBL, BI, xG, wG = Legendre_Gauss_Basis(3)",
    "output": "Nós ≈ [-0.7746, 0.0, 0.7746]\nPesos ≈ [0.5556, 0.8889, 0.5556]\nPrimeiras linhas de BL/BI incluídas"
  },
  {
    "id": "legendre-lobatto-basis",
    "title": "Legendre_Lobatto_Basis",
    "description": "Nós e pesos de Gauss-Lobatto com a respectiva matriz de base.",
    "code": "using SpectralLib\nxLL, wGL, P = Legendre_Lobatto_Basis(4)",
    "output": "Nós ≈ [-1.0, -0.4472, 0.4472, 1.0]\nPesos ≈ [0.1667, 0.8333, 0.8333, 0.1667]\nMatriz P montada para ordem 4"
  },
  {
    "id": "legendre-radau-left",
    "title": "Legendre_Radau_BasisL",
    "description": "Base de Radau com nó fixo em -1.",
    "code": "using SpectralLib\nxRL, wRL, RL = Legendre_Radau_BasisL(4)",
    "output": "Nós ≈ [-1.0, -0.6547, 0.1424, 0.7560]\nPesos ≈ [0.1, 0.5444, 0.4947, 0.8609]\nBase RL calculada"
  },
  {
    "id": "legendre-radau-right",
    "title": "Legendre_Radau_BasisR",
    "description": "Base de Radau com nó fixo em 1.",
    "code": "using SpectralLib\nxRR, wRR, RR = Legendre_Radau_BasisR(4)",
    "output": "Nós ≈ [-0.7560, -0.1424, 0.6547, 1.0]\nPesos ≈ [0.8609, 0.4947, 0.5444, 0.1]\nBase RR calculada"
  },
  {
    "id": "legendre-eval",
    "title": "eval_legendre",
    "description": "Avalia polinômios de Legendre em pontos fornecidos.",
    "code": "using SpectralLib\nvals = eval_legendre(3, [-0.5, 0.0, 0.5])",
    "output": "eval_legendre(5, [-1, -0.5, 0, 0.5, 1]) -> matriz 6×5 com P₀..P₅"
  },
  {
    "id": "legendre-quad-types",
    "title": "GaussQuadTypes",
    "description": "Quadratura genérica em um intervalo com vários tipos de nós.",
    "code": "using SpectralLib\nz, w = GaussQuadTypes(-1.0, 1.0, 3, 2)",
    "output": "GaussQuadTypes(-1,1,4,tipo=3) -> z≈[-1.0,-0.5,0.5,1.0]\nw≈[0.0833,0.4167,0.4167,0.0833]"
  },
  {
    "id": "legendre-ds",
    "title": "DS_Legendre",
    "description": "Matriz modal de diferenciação para Legendre.",
    "code": "using SpectralLib\nD = DS_Legendre(3)",
    "output": "DS_Legendre(3) 4×4:\n[[0.0, 3.0, -3.0, 0.0],\n [-0.75, 0.0, 0.75, 0.0],\n [0.0, -0.75, 0.0, 0.75],\n [0.0, 3.0, -3.0, 0.0]]"
  },
  {
    "id": "legendre-basis-matrix",
    "title": "Base_Legendre_n",
    "description": "Matriz de Vandermonde de Legendre até ordem n.",
    "code": "using SpectralLib\nx = [-1.0, 0.0, 1.0]\nB = Base_Legendre_n(3, x)",
    "output": "Base_Legendre_n(3, [-1,0,1]) ->\n[[1.0,-1.0,1.0,-1.0],\n [1.0,0.0,-0.5,0.0],\n [1.0,1.0,1.0,1.0]]"
  },
  {
    "id": "legendre-mprod",
    "title": "M_Prod_Legendre",
    "description": "Matriz de multiplicação espectral usando coeficientes de Legendre.",
    "code": "using SpectralLib\nah = [1.0, 0.5, 0.25]\nM = M_Prod_Legendre(ah)",
    "output": "M_Prod_Legendre(ah=[1.0,0.3,0.1]) -> matriz 3×3 com termos de multiplicação"
  },
  {
    "id": "legendre-integration",
    "title": "JS_Leg",
    "description": "Matriz operacional de integração para coeficientes de Legendre.",
    "code": "using SpectralLib\nJ = JS_Leg(3)",
    "output": "JS_Leg(3) 4×4 integração modal"
  },
  {
    "id": "fourier-ds",
    "title": "DS_Fourier",
    "description": "Matriz modal de diferenciação para coeficientes reais de Fourier.",
    "code": "using SpectralLib\nD = DS_Fourier(2)",
    "output": "D1 e D2 para N=8: D1[0:3,0:3] ≈ [[0,0.7854,0,-0.7854],[ -0.7854,0,0.7854,0],[0,-0.7854,0,0.7854]]"
  },
  {
    "id": "fourier-fourdif",
    "title": "fourdif",
    "description": "Matriz de diferenciação espectral para pontos periódicos.",
    "code": "using SpectralLib\nx, D = fourdif(5, 1)",
    "output": "fourdif(9,1) -> x ≈ [0,0.6981,1.3963,2.0944,2.7925,3.4907,4.1888,4.8869,5.5851]\nD1 circulante"
  },
  {
    "id": "fourier-multi-diff",
    "title": "Multi_Diff_Mat",
    "description": "Coleção de matrizes de diferenciação até uma ordem máxima.",
    "code": "using SpectralLib\nx, mats = Multi_Diff_Mat(5, 2)",
    "output": "D1, D2, D3 para N=6 empilhadas (derivadas 1..3)"
  },
  {
    "id": "fourier-barycentric",
    "title": "Bary_Trig_Mat",
    "description": "Interpolação trigonométrica em pontos arbitrários.",
    "code": "using SpectralLib\nfk = [0.0, 1.0, 0.0, -1.0]\nx = [0.1, 2.0]\nt, P = Bary_Trig_Mat(fk, x)",
    "output": "TrigBarycentricInterp(n=6) -> valores interpolados ≈ [-0.0,0.8660,0.8660,-0.0,-0.8660,-0.8660]"
  },
  {
    "id": "fourier-integration",
    "title": "JS_Fourier",
    "description": "Matriz de integração modal para séries reais de Fourier.",
    "code": "using SpectralLib\nI = JS_Fourier(2)",
    "output": "JS_Fourier(5) -> matriz 5×5 antiderivada trigonométrica"
  },
  {
    "id": "fourier-quad-function",
    "title": "fourier_quad (função)",
    "description": "Integra uma função periódica usando quadratura espectral.",
    "code": "using SpectralLib\nintegral = fourier_quad(x -> sin(x), 9)",
    "output": "Trig_Quad(f=sin,0,2π,8) ≈ 2.0"
  },
  {
    "id": "fourier-quad-data",
    "title": "fourier_quad (vetor)",
    "description": "Integra dados discretos igualmente espaçados.",
    "code": "using SpectralLib\nsamples = sin.(range(0, stop=2pi, length=9))\nvalue = fourier_quad(samples)",
    "output": "Trig_Quad_data(sin.(nodes), 0, 2π, 8) ≈ 2.0"
  },
  {
    "id": "fourier-real-fft",
    "title": "real_fft / real_ifft",
    "description": "Transformada rápida real e inversa para vetores.",
    "code": "using SpectralLib\nvals = [0.0, 1.0, 0.0, -1.0]\nspectral = real_fft(vals)\nrecovered = real_ifft(spectral)",
    "output": "real_fft(sin.(x)) para x em [0,2π) com 8 pontos -> coeficientes principais ≈ [0,0,-4.0,0,0]"
  },
  {
    "id": "generic-barycentric",
    "title": "Bary_Interp",
    "description": "Interpolação polinomial baricêntrica sobre pontos arbitrários.",
    "code": "using SpectralLib\nxk = [-1.0, 0.0, 1.0]\nfk = [1.0, 0.0, 1.0]\nxnew = [-0.5, 0.5]\nfnew, P = Bary_Interp(xk, fk, xnew)",
    "output": "barycentric_interpolation([0,1,2],[1,3,2],[0.5,1.5]) -> [2.0,2.5]"
  },
  {
    "id": "generic-diff-matrix",
    "title": "Generalized_Diff_Mat",
    "description": "Matriz de diferenciação para pontos não uniformes.",
    "code": "using SpectralLib\nxs = [-1.0, -0.2, 0.4, 1.0]\nD = Generalized_Diff_Mat(xs)",
    "output": "differentiation_matrix([-1,-0.5,0,0.5,1]) -> matriz 5×5 com derivadas de base racional"
  }
]
